src/harness/reference_models/dpa/dpa_mgr.py

```diff
@@ -135,27 +137,40 @@ def __init__(self, protected_points,
                beamwidth=3,
                azimuth_range=(0, 360),
                freq_ranges_mhz=[(3550, 3650)],
-               catb_neighbor_dist=ml.CAT_B_NBRHD_DIST_DEFAULT):
+               catb_neighbor_dist=ml.CAT_B_NBRHD_DIST_DEFAULT,
+               dpa_type='esc'):
     """Initialize the DPA attributes."""
     self.protected_points = protected_points
     self.threshold = threshold
     self.radar_height = radar_height
     self.azimuth_range = azimuth_range
     self.beamwidth = beamwidth
-    self.channels = GetDpaProtectedChannels(freq_ranges_mhz)
     self.catb_neighbor_dist = catb_neighbor_dist
+    self.dpa_type = dpa_type
+    self.channels = None
     self._grants = []
     self.has_th_grants = False
+    self.ResetFreqRange(freq_ranges_mhz)
     self.ResetLists()
 
+  def __str__(self):
+    """Returns the DPA str."""
+    return ('Dpa(protected_points=%r, threshold=%.1f, radar_height=%.1f,'
+            'beamwidth=%.1f, azimuth_range=%r, channels=%r,'
+            'catb_neighbor_dist=%.1f, dpa_type=%r)' % (
+                self.protected_points, self.threshold, self.radar_height,
+                self.beamwidth, self.azimuth_range, self.channels,
+                self.catb_neighbor_dist, self.dpa_type))
+
   def ResetFreqRange(self, freq_ranges_mhz):
     """Reset the frequency ranges of the DPA.
 
     Args:
       freq_ranges_mhz: The protection frequencies (MHz) as a list of tuple
         (freq_min_mhz, freq_max_mhz) of the DPA protected frequency ranges.
     """
-    channels = GetDpaProtectedChannels(freq_ranges_mhz)
+    channels = GetDpaProtectedChannels(freq_ranges_mhz,
+                                       is_portal_dpa=(self.dpa_type=='portal'))
     if channels != self.channels:
       self.ResetLists()
     self.channels = channels
```

Comment:
Could you please explain why "channels" is compared with "self.channels", and if True, self.ResetLists() is called? And then self.ResetLists() is called again (line 154)?

Reply:
Also, the description of ResetLists() should be """Reset move list and neighbor list."""

Reply:
We need to reset if the state of the DPA change, ie grants are changed or channel are changed.
I agree that all this is not fully hermetic and could be improved. If I change the radar height, the list will become invalid as well, but currently we dont manage that, as we do not expect that someone would do such thing.

The channel list change may happen however the way the tests are written, I am not fully sure anymore. And the grants too.
Perhaps I should revisit all this to be a pure calculation class without state keeping - just returning the lists as standalone objects. But I guess it is a bit late now to do massive architectural change.

Reply:
Done for the description.